src/notebooks/flatten_and_explode.ipynb

```diff
@@ -0,0 +1,524 @@
+       "arguments": {},
+       "data": "<div class=\"ansiout\"></div>",
+       "datasetInfos": [],
+       "metadata": {},
+       "removedWidgets": [],
+       "type": "html"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "'''\n",
+    "Transforms a spark dataframe into an \"exploded\" data frame. To explode a data frame \n",
+    "means to add a new row containing identical identifying information to a base row \n",
+    "for each array-element encountered in columns whose fields are structs. The result\n",
+    "is returned as a dataframe.\n",
+    "'''\n",
+    "def explode_arrays(df):\n",
+    "    flat_cols = [field.name for field in df.schema.fields if type(field.dataType) != ArrayType]\n",
+    "    struct_cols = [field.name for field in df.schema.fields if type(field.dataType) == ArrayType]\n",
+    "    \n",
+    "    # Add new rows for each array element, then select and join with columns\n",
+    "    # that are already flat\n",
+    "    exploded_df = df.withColumn('vals', explode_outer(arrays_zip(*struct_cols))) \\\n",
+    "           .select(*flat_cols,'vals.*') \\\n",
+    "           .fillna('', subset=struct_cols)\n",
+    "    return exploded_df"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "inputWidgets": {},
+     "nuid": "58b88ca4-f51d-4ec2-9852-e0af3628ca6e",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/html": [
+       "<style scoped>\n",
+       "  .ansiout {\n",
+       "    display: block;\n",
+       "    unicode-bidi: embed;\n",
+       "    white-space: pre-wrap;\n",
+       "    word-wrap: break-word;\n",
+       "    word-break: break-all;\n",
+       "    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n",
+       "    font-size: 13px;\n",
+       "    color: #555;\n",
+       "    margin-left: 4px;\n",
+       "    line-height: 19px;\n",
+       "  }\n",
+       "</style>\n",
+       "<div class=\"ansiout\"></div>"
+      ]
+     },
+     "metadata": {
+      "application/vnd.databricks.v1+output": {
+       "addedWidgets": {},
+       "arguments": {},
+       "data": "<div class=\"ansiout\"></div>",
+       "datasetInfos": [],
+       "metadata": {},
+       "removedWidgets": [],
+       "type": "html"
+      }
+     },
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "'''\n",
+    "A wrapper function to perform full flattening of an input spark data frame.\n",
+    "Struct columns are separated out into multiple columns, and then the intermediate\n",
+    "expanded data frame is exploded so that each new element encountered generates a \n",
+    "new row.\n",
+    "'''\n",
+    "def flatten_df(dfflat):\n",
+    "    while len([field.name for field in dfflat.schema.fields if type(field.dataType) == StructType or type(field.dataType) == ArrayType ]) !=0 :\n",
```

Comment:
You can make this a little cleaner with the following:

```
while len([field.name for field in dfflat.schema.fields if type(field.dataType) in [StructType, ArrayType]]) != 0:
```

Additionally, while not really an issue here, I think it's best to get into the habit of being explicit about what conditions you're testing for. Here, you've said that as long as the length does not equal 0 then lines 274 and 275 should run, which has an implicit understanding that `len` can only be positive. Implementing these implicit assumptions can lead to unforeseen outcomes and I'd recommend trying to avoid them where possible. To that end, I'd rewrite this to be:

```
while len([field.name for field in dfflat.schema.fields if type(field.dataType) in [StructType, ArrayType]]) > 0:
```